### PR TITLE
perf: Improve the performance of the provisioner

### PR DESCRIPTION
**Pull request by stijndehaes**
In support of: https://github.com/aws/karpenter/issues/3565

**Description**

This changes improves the performance of the provisioner. We saw high CPU usage, using pprof I found out that a lot of CPU cycles are wasted on the Union function on the string set.

Instead of using union I inserted all the new strings in place in the set, this decreased CPU usage by up to 50%.

**How was this change tested?**

I tested it out by building karpenter from source and deploying on a cluster with high CPU usage. You can see the result here. The green line is the before and the yellow after the change:
image

I also have 2 pprof profiles, one before the change you can find the pdf here:
[profile-old.pdf](https://github.com/aws/karpenter-core/files/10942733/profile-old.pdf)

And one after the change:
[profile-improved.pdf](https://github.com/aws/karpenter-core/files/10942739/profile-improved.pdf)

You can clearly see that the hot code path has changed 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/provisioning/provisioner.go [hunk 1]
```diff
@@ -227,13 +227,21 @@ func (p *Provisioner) NewScheduler(ctx context.Context, pods []*v1.Pod, stateNod
 			requirements.Add(instanceType.Requirements.Values()...)
 
 			for key, requirement := range requirements {
-				domains[key] = domains[key].Union(sets.NewString(requirement.Values()...))
+				if domains[key] == nil {
```
Comment (jonathan-innis) [inline pkg/controllers/provisioning/provisioner.go:230]: I can see someone coming along later and asking why this was done. Do you mind commenting (linking the issue?) as well as giving a brief explanation of the performance improvement here?

**Review by jonathan-innis — COMMENTED** [review]
This is awesome! Just one thought on giving an explanation

**Comment by ellistarn** [comment]
Can you run the scheduling benchmark before and after? Would love to see the results.

File: pkg/controllers/provisioning/provisioner.go (hunk 1 above)
Comment (stijndehaes) [reply pkg/controllers/provisioning/provisioner.go:234]: Added the requested comments, let me know if this is clear

**Comment by stijndehaes** [comment]
> Can you run the scheduling benchmark before and after? Would love to see the results.

I would love to, but I can not find it. Can you help me with that?

**Comment by ellistarn** [comment]
> Looks rather similar to me

This is what I get for reviewing code on my phone. Apologies -- this provisioner code is outside of the scope of the benchmark 🤦 . I'm honestly shocked to see this amount of performance hit from this piece of code, given that it's essentially initialization logic. Do you have a massive number of provisioners, or something?

**Comment by stijndehaes** [comment]
> > Looks rather similar to me
> 
> This is what I get for reviewing code on my phone. Apologies -- this provisioner code is outside of the scope of the benchmark 🤦 . I'm honestly shocked to see this amount of performance hit from this piece of code, given that it's essentially initialization logic. Do you have a massive number of provisioners, or something?

@ellistarn we only use 4 provisioners, 3 of them running emptiness, and one using consolidation. We run from 50 to 800 nodes, and are constantly scheduling batch jobs
You can find more info here: https://github.com/aws/karpenter/issues/3565

You can also see that most of the time of the CPU goes to the sets.Union, because there are many copies of data happening here. In general in the code there are many copies happening resulting in a lot of CPU cycles. I realise you guys have been focussing on features, but I would like to help you guys improve the performance :) Maybe I can try to create some benchmarks that show the same performance issues I am seeing?

**Comment by ellistarn** [comment]
> I realise you guys have been focussing on features

Agree. We're shifting gears as we approach v1.

> but I would like to help you guys improve the performance :) Maybe I can try to create some benchmarks that show the same performance issues I am seeing?

Fantastic. Can you work with @jonathan-innis to help define this? Are you thinking about code benchmarks or real test workloads? We need to build these benchmarks into our e2e suites @spring1843

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Pull request merged**